Introduce basic transactions API

kandrio opened the pull request:
Everything is summarized in our commit messages and on the following issue of our team:
- https://github.com/Tribler/tribler/issues/7293

OrestisKan commented on detoks/src/main/java/nl/tudelft/trustchain/detoks/DeToksCommunity.kt:
```diff
@@ -2,13 +2,14 @@ package nl.tudelft.trustchain.detoks
-class DeToksCommunity(private val context: Context) : Community() {
+class DeToksCommunity(
+    private val context: Context
+    ) : TransactionEngine("c86a7db45eb3563ae047639817baec4db2bc7c25") {
```
I understand that serviceId was hardcoded in the code but it seems to me that it should not be, maybe use some environment file or something to hide it?

OrestisKan commented on detoks/src/main/java/nl/tudelft/trustchain/detoks/LightweightTxApp.kt:
```diff
@@ -0,0 +1,114 @@
+    private fun createCommunity(): OverlayConfiguration<TransactionEngine> {
+        val randomWalk = RandomWalk.Factory(timeout = 3.0, peers = 20)
+        return OverlayConfiguration(
+            TransactionEngine.Factory("c86a7db45eb3563ae047639817baec4db2bc7c25"),
```
Again hardcoded

OrestisKan commented on detoks/src/main/java/nl/tudelft/trustchain/detoks/LightweightTxApp.kt:
```diff
@@ -0,0 +1,114 @@
+            while (true) {
+                val overlay = ipv8.overlays.values.toList()[0]
+                printPeersInfo(overlay)
+                println("===")
```
remove

kandrio replied:
Thanks for the comment. Could you be a bit more specific here? Are you referring to the entire `while (true)` statement? Or to the `printPeersInfo(overlay)` part?

OrestisKan commented on detoks/src/main/java/nl/tudelft/trustchain/detoks/LightweightTxApp.kt:
```diff
@@ -0,0 +1,114 @@
+
+        val numberOfTransactions = 1
+        scope.launch {
+            while (true) {
```
Bad practice

OrestisKan commented on detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    fun sendTransaction(blockBuilder: BlockBuilder, peer: Peer?, encrypt: Boolean = false) {
+        println("sending block...")
```
Debug print ? Use logger if needed

kandrio replied:
I agree. I will use a logger from the `mu.KotlinLogging` library

OrestisKan commented on detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt:
```diff
@@ -0,0 +1,117 @@
+                e.printStackTrace()
+            }
+        } else {
+            println("Received unknown message $msgId from $sourceAddress")
```
Use logger

OrestisKan commented on detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    private fun msgIdFixer(msgid: Int): Int{
+        @Suppress("DEPRECATION")
```
Not a good practice, there should be some other approach for the same action

kandrio replied:
I agree. The problem was that we used `MessageId` flags (e.g: `HALF_BLOCK`, `HALF_BLOCK_ENCRYPTED` etc...) that were larger than a `Ubyte`. I will use smaller values as flags and I will remove the `msgIdFixer`.

kandrio commented:
Thank you very much for the comments. We will address them as soon as
possible.

Στις Τρί, 28 Μαρ 2023, 14:15 ο χρήστης Orestis Kanaris <
***@***.***> έγραψε:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In detoks/src/main/java/nl/tudelft/trustchain/detoks/DeToksCommunity.kt
> <https://github.com/Tribler/trustchain-superapp/pull/145#discussion_r1150422764>
> :
>
> >  import nl.tudelft.ipv8.Overlay
>  import nl.tudelft.ipv8.Peer
>  import nl.tudelft.ipv8.messaging.Packet
>
>
> -class DeToksCommunity(private val context: Context) : Community() {
> +class DeToksCommunity(
> +    private val context: Context
> +    ) : TransactionEngine("c86a7db45eb3563ae047639817baec4db2bc7c25") {
>
> I understand that serviceId was hardcoded in the code but it seems to me
> that it should not be, maybe use some environment file or something to hide
> it?
> ------------------------------
>
> In detoks/src/main/java/nl/tudelft/trustchain/detoks/LightweightTxApp.kt
> <https://github.com/Tribler/trustchain-superapp/pull/145#discussion_r1150424399>
> :
>
> > +import nl.tudelft.ipv8.sqldelight.Database
> +import java.net.InetAddress
> +import java.util.*
> +import kotlin.math.roundToInt
> +
> +class DemoTransactionApp {
> +    private val scope = CoroutineScope(Dispatchers.Default)
> +
> +    fun run() {
> +        startIpv8()
> +    }
> +
> +    private fun createCommunity(): OverlayConfiguration<TransactionEngine> {
> +        val randomWalk = RandomWalk.Factory(timeout = 3.0, peers = 20)
> +        return OverlayConfiguration(
> +            TransactionEngine.Factory("c86a7db45eb3563ae047639817baec4db2bc7c25"),
>
> Again hardcoded
> ------------------------------
>
> In detoks/src/main/java/nl/tudelft/trustchain/detoks/LightweightTxApp.kt
> <https://github.com/Tribler/trustchain-superapp/pull/145#discussion_r1150424866>
> :
>
> > +        ), walkerInterval = 1.0)
> +
> +        // storage
> +        val driver: SqlDriver = JdbcSqliteDriver(JdbcSqliteDriver.IN_MEMORY)
> +        Database.Schema.create(driver)
> +        val store = TrustChainSQLiteStore(Database(driver))
> +
> +        val ipv8 = IPv8(endpoint, config, myPeer)
> +        ipv8.start()
> +
> +        val numberOfTransactions = 1
> +        scope.launch {
> +            while (true) {
> +                val overlay = ipv8.overlays.values.toList()[0]
> +                printPeersInfo(overlay)
> +                println("===")
>
> remove
> ------------------------------
>
> In detoks/src/main/java/nl/tudelft/trustchain/detoks/LightweightTxApp.kt
> <https://github.com/Tribler/trustchain-superapp/pull/145#discussion_r1150425239>
> :
>
> > +
> +        val config = IPv8Configuration(overlays = listOf(
> +            createCommunity()
> +        ), walkerInterval = 1.0)
> +
> +        // storage
> +        val driver: SqlDriver = JdbcSqliteDriver(JdbcSqliteDriver.IN_MEMORY)
> +        Database.Schema.create(driver)
> +        val store = TrustChainSQLiteStore(Database(driver))
> +
> +        val ipv8 = IPv8(endpoint, config, myPeer)
> +        ipv8.start()
> +
> +        val numberOfTransactions = 1
> +        scope.launch {
> +            while (true) {
>
> Bad practice
> ------------------------------
>
> In detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt
> <https://github.com/Tribler/trustchain-superapp/pull/145#discussion_r1150427546>
> :
>
> > +    object MessageId {
> +        const val HALF_BLOCK = 4242
> +        const val HALF_BLOCK_ENCRYPTED = 4243
> +        const val HALF_BLOCK_BROADCAST = 424242
> +        const val HALF_BLOCK_BROADCAST_ENCRYPTED = 424243
> +    }
> +
> +    init {
> +        messageHandlers[msgIdFixer(MessageId.HALF_BLOCK)] = ::onHalfBlockPacket
> +        messageHandlers[msgIdFixer(MessageId.HALF_BLOCK_BROADCAST)] = ::onHalfBlockBroadcastPacket
> +        messageHandlers[msgIdFixer(MessageId.HALF_BLOCK_ENCRYPTED)] = ::onHalfBlockPacket
> +        messageHandlers[msgIdFixer(MessageId.HALF_BLOCK_BROADCAST_ENCRYPTED)] = ::onHalfBlockBroadcastPacket
> +    }
> +
> +    fun sendTransaction(blockBuilder: BlockBuilder, peer: Peer?, encrypt: Boolean = false) {
> +        println("sending block...")
>
> Debug print ? Use logger if needed
> ------------------------------
>
> In detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt
> <https://github.com/Tribler/trustchain-superapp/pull/145#discussion_r1150427826>
> :
>
> > +            // logger.debug("prefix not matching")
> +            return
> +        }
> +
> +        val msgId = data[prefix.size].toUByte().toInt()
> +
> +        val handler = messageHandlers[msgId]
> +
> +        if (handler != null) {
> +            try {
> +                handler(packet)
> +            } catch (e: Exception) {
> +                e.printStackTrace()
> +            }
> +        } else {
> +            println("Received unknown message $msgId from $sourceAddress")
>
> Use logger
> ------------------------------
>
> In detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt
> <https://github.com/Tribler/trustchain-superapp/pull/145#discussion_r1150428117>
> :
>
> > +
> +        val handler = messageHandlers[msgId]
> +
> +        if (handler != null) {
> +            try {
> +                handler(packet)
> +            } catch (e: Exception) {
> +                e.printStackTrace()
> +            }
> +        } else {
> +            println("Received unknown message $msgId from $sourceAddress")
> +        }
> +    }
> +
> +    private fun msgIdFixer(msgid: Int): Int{
> +        @Suppress("DEPRECATION")
>
> Not a good practice, there should be some other approach for the same
> action
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Tribler/trustchain-superapp/pull/145#pullrequestreview-1360796663>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AL23YTVR5VSSPJ5LPONJ2WTW6LB5PANCNFSM6AAAAAAWJIIOEA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

kandrio commented:
@OrestisKan We updated our PR and addressed all of your comments. We decided to drop the `LightweightTxApp` from the PR. We'll keep it internally for testing and won't push it as a "production-ready" code. 

Feel free to take a look again at our branch. We're to clarify any questions you may have

synctext commented:
@OrestisKan is a bit overloaded, as nobody made the ```Week 6``` deadline for Prototype I. Review will come asap.

OrestisKan reviewed (changes requested):
Code seems alright, but I'm not sure how to work the app, maybe you can guide me ? 
Also fix the merge conflicts please.

OrestisKan reviewed (approved):
Seems alright!
